### DISCUSSION TITLE
How about add event government to Numaflow

### DISCUSSION THREAD

**Discussion by JasonChen86899** [numaproj/numaflow#904] [Ideas]
Earlier I have designed an spec protocol of event government. That is my experience of building a pipeline system like Numaflow
So I want to add this config of event government to Numaflow if this spec protocol match Numaflow

**Comment by vigith** [comment]
What do you mean by event government? Could you please provide some details on this?

**Reply by JasonChen86899** [reply]
we may know some service government strategies such as rate_limiit, fault_tolerance, circuit_breaker, Retry. 
In EDA system or streaming data processing sense, we also need some strategies to make our system stable.
But it is different from service government because of the infra such as MQ and async model, so I has designed a spec config that includes event government strategies.
It includes event source sink channel runtime and router config

I will copy my doc url here, you can review it

**Reply by vigith** [reply]
Please share so that I can read about it.

**Reply by JasonChen86899** [reply]
ok, this is the doc [Event Government](https://docs.qq.com/doc/DVEpncUxIc0p3V21K), please review it

In this doc, there is some other proto `io.opensergo.proto` which is referenced. You can ignore it. They are common service government strategies. we can replace it with ourself If so.Just review other event proto model

**Reply by vigith** [reply]
I read the doc; we support most of the strategies but not all. Eg, we do have inbuilt retries, routing capabilities, exactly-once, DLQ via routing, etc. We do not have inbuilt triggers, but we can Numaflow couple with other triggering solutions like K8s cron.

Though the spec looks different from Numaflow, Numaflow supports these features inherently as these are quite important for any stream processing system.

**Reply by JasonChen86899** [reply]
Thx,most of them is same, but triggers in my doc is like UDF in Numa.  This is my Summary of trigger and sink
1. knative eventing use trigger to be the consumer of mq. sink is defined by self in trigger yaml file
2. argo event use senor to be consumer and trigger to be next input. sink is merged into trigger and has no its separate definitions

so I define trigger in this way. 
By the way, I want to summarize a sepc pb of event. so hope you to review and give me some suggestion  about it. we can propose an common standard protocol together and then promote it to more eventing framework :)

**Reply by vigith** [reply]
Numaflow can support any type of event because it has inbuilt support for [user-defined-source-transformer](https://numaflow.numaproj.io/user-guide/sources/transformer/overview/). We can make a builtin for a CNCF native event spec if there is enough adoption.

**Reply by JasonChen86899** [reply]
CNCF native event spec that sounds excited!  I think we can set up a development talk group and discuss this matter. It is also convenient to communicate and bring in more friends from other community project such as knative eventing
Slack or Discard?

**Reply by vigith** [reply]
sounds good to me. @syayi could you please take a look from the produce perspective?

**Reply by vigith** [reply]
@JasonChen86899 we do have a slack channel, we can discuss it there?

**Reply by syayi** [reply]
Sure, @vigith! 

@JasonChen86899 We would love to learn your use case 

Here is the [Slack community](https://numaproj.slack.com/join/shared_invite/zt-19svuv47m-YKHhsQ~~KK9mBv1E7pNzfg#/shared-invite/email), and looking forward to connecting with you.
